node/src/ledger/ledger_service.rs

```diff
@@ -72,12 +72,98 @@ fn merkle_root(mask: &mut Mask) -> LedgerHash {
     MinaBaseLedgerHash0StableV1(mask.merkle_root().into()).into()
 }
 
+/// Indexing `StagedLedger` both by their "merkle root hash" and their "staged ledger hash"
+#[derive(Default)]
+struct StagedLedgersStorage {
+    /// 1 merkle root hash can refer to 1 or more `StagedLedger`
+    by_merkle_root_hash: BTreeMap<LedgerHash, Vec<Arc<MinaBaseStagedLedgerHashStableV1>>>,
+    staged_ledgers: BTreeMap<Arc<MinaBaseStagedLedgerHashStableV1>, StagedLedger>,
```

Comment:
Not super important, but what do you think of having a `v2::StagingLedgerHash` alias like in my branch?